### ISSUE TITLE
Enable monitor mode on hammerheadmon nougat ?

### ISSUE THREAD

**Issue by malokoum** [offensive-security/kali-nethunter#713]
**Device:** 
Nexus 5

**OS version (KitKat/Lollipop/Marshmallow/Nougat):**
CM14 Nougat

**Built from repo (date and build command) or downloaded from website (links):** 
https://github.com/offensive-security/kali-nethunter
python2 build.py -d hammerheadmon -n --rootfs full

**Output of cat /proc/version in adb shell or NetHunter Terminal:**
Linux version 3.4.0-g1262232-dirty (autintech@autintech) (gcc version 4.7 (GCC) ) #2 SMP PREEMPT Sun Dec 11 16:08:30 EST 2016

Hi,

How to configure WiFi adaptor wlan0 to monitor mode ?

Thx

**Comment by akrolla** [comment]
Source monstart
Source monstop I believe its on the wiki

**Comment by akrolla** [comment]
Source monstart-nh
Source monstop-nh

**Comment by AutinDroid** [comment]
@malokoum Hi, I developed the kernel image for hammerhead-nougat. You will have to download this: https://github.com/AutinDroid/HammerheadMon-Kernel/releases/download/v0.1/nexmon.zip 
Extract the contents of the folder in the zip to /system/xbin. Don't forget to install firmware via the nexmon app (play store). To start monitor mode, do `source monstart-nh` and to stop do `source monstop-nh`.

**Comment by jcadduono** [comment]
@AutinDroid might be a good idea to include that in with the kernel, ex. devices/nougat/hammerheadmon/system/xbin/*

**Comment by AutinDroid** [comment]
OK will do that when I get back from vacation.

**Comment by jcadduono** [comment]
2slow i win, done

**Comment by AutinDroid** [comment]
:( I did do the system/xbin thing a while back: https://github.com/AutinDroid/HammerheadMon-Kernel/releases

**Comment by jcadduono** [comment]
if u check the zip its not actually in there :P i just added system folder inclusion to the installer a few hours ago heheheh

**Comment by AutinDroid** [comment]
@jcadduono sniff sniff... I admit defeat ;(

**Comment by jcadduono** [comment]
https://build.nethunter.com/nightly/3.15.3-20161221-0941/ :) good to go

**Comment by AutinDroid** [comment]
OK I'll be downloading while crying :)

**Comment by akrolla** [comment]
Perhaps this could fix my issue toooooo 
Docu have a revoery image of it yet that I could flash since mines a caf right now I think that is where my problem lies

**Comment by akrolla** [comment]
Do you*

**Comment by AutinDroid** [comment]
@akrolla I'll build an hammerheadcafcmmon-nougat image (wow long name, wait till hammerheadcafcmmon-nougat is created) unless if @jcaddunno beats me to it... Again.

**Comment by akrolla** [comment]
Jcad ur a beast dude might I inquire what programming languages do you know bet you know a few don't you?

**Comment by akrolla** [comment]
Lmao yea that's what my problem is

**Comment by akrolla** [comment]
Now I can close my nexus crashing issue

**Comment by akrolla** [comment]
As tears of joy sweep my face because youre right -_- I some of the names of the kernel updates are a little confusing for the hammerhead platform

**Comment by AutinDroid** [comment]
Lol
Wait till I create hammerheadmonelementalxoverclcked2.7ghzcafcmmon

**Comment by akrolla** [comment]
Hahaha should make one superkalifragileroutersnethunteros

**Comment by akrolla** [comment]
Can't wait till my mobile development course in college I can take it next semester I can't freaking wait then I can at least attempt to help more than I can now-_-

**Comment by AutinDroid** [comment]
@akrolla Great, I can't wait till I get into college!

**Comment by malokoum** [comment]
Hi,
Good job @AutinDroid  @jcadduono 

https://github.com/offensive-security/nethunter-devices/tree/master/nougat/hammerheadmon/system/xbin
Just change the variable LD_PRELOAD in monstart-nh so that the lib is loaded...
sed -i '/xbin/lib/' monstart-nh

**Comment by AutinDroid** [comment]
@jcadduono you forgot to include libfakeioctl.so and libfakeioctlkali.so (or whatever it's called). grr....

**Comment by jcadduono** [comment]
https://github.com/offensive-security/nethunter-devices/tree/master/nougat/hammerheadmon/system/lib ? it's there

**Comment by jcadduono** [comment]
oh why is the LD_PRELOAD lib location in xbin? that doesnt make any sense

**Comment by AutinDroid** [comment]
@jcadduono My mistake. NethunterOS has the libs in system/xbin, so I thought we'd have to.

**Comment by dverlik** [comment]
@AutinDroid Sorry to bother, but it seems that in my case the LD_PRELOAD isn't finding the .so file. I've placed the libfakeioctl.so in multiple places:

- /system/lib/libfakeioctl.so
- /system/xbin/libfakeioctl.so
- /system/xbin/lib/libfakeioctl.so
- /usr/lib/arm-linux-gnueabihf/libfakeroot/libfakeioctl.so
- /usr/lib/arm-linux-gnueabihf/libfakeioctl.so

But every time I change the LD_PRELOAD to "libfakeioctl.so" it just gives me that error:

> ERROR: ld.so: object 'libfakeioctl.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
I've tried to fill in the LD_PRELOAD the path to the lib, but it still fails.

I have Nexus 5 with kali nethunter 3 installed in multirom mode, using the multirom app & multirom TWRP recovery. Can you please suggest something to do to fix this?

**Comment by AutinDroid** [comment]
@dverlik You are using the hammerheadmon-nougat package, right? That package installs everything in the right place, and the loaded .So is libfacekaliioctl.so (after installing the kernel package, you can run "source monstart-nh" to preload it in chroot. Marshmallow hammerheadmon is broken, and by nethunter 3, do you mean the official installer 0_o (that thing is deprecated).

**Comment by dverlik** [comment]
@AutinDroid Well, looks like I'm using and old package - I downloaded the .zip from [this offsec page](https://www.offensive-security.com/kali-linux-nethunter-download/) - the zip for my nexus 5 with lollipop. Where can I download the non-deprecated version?

**Comment by AutinDroid** [comment]
@dverlik the correct page is https://build.nethunter.com/nightly/
Also, you need to install nougat as the kernel is hammerheadmon-nougat (can find in page). After flashing page, flash a full or minimal kalifs (armhf). To activate monitor mode, type "monstart-nh in a Kali terminal".

**Comment by dverlik** [comment]
@AutinDroid Thank you! So, the correct order of installation would be:

1. Install OTA marshmallow for hammerhead - the original android marshmallow for nexus 5
2. Flash `kernel-nethunter-hammerhead-nougat-3.15.4-20170202-1604.zip` (the custom kernel)
3. Flash `SuperSU.zip` (what would be the correct version?) as the kernel now is rootable
4. Flash `kernel-nethunter-hammerhead-marshmallow-3.15.4-20170202-1604.zip`
5. Download & install chroot via the Nethunter app

Again, thank you a lot for helping!

**Comment by AutinDroid** [comment]
@dverlik err... That still is wrong. You replacing the nougat kernel with a marshmallow kernel will do nothing, except install the marshmallow kernel. You need a Nougat ROM, can't use a marshmallow, for the nexmon capabilities. Flash ONLY the hammerheadmon-nougat kernel and the Kali nethunter kalifs (generic armhf), both found in the page I gave you. You MUST be using a nougat ROM, flashing a marshmallow kernel on nougat can hurt your phone, so please don't. After the proper kernel and kalifs are flashed, download the nexmon app from play store and install firmware. You can run "source monstart-nh" in Kali terminal to activate monitor mode.

**Comment by akrolla** [comment]
https://gist.github.com/binkybear/8f0f4eee292d5263372f47c3c3edd08e

**Comment by dverlik** [comment]
@AutinDroid So, that would correct:
1. Create a new rom in multirom recovery.
2. Flash the `kernel-nethunter-hammerhead-nougat-3.15.4-20170202-1604.zip`
3. Flash `kernel-nethunter-generic-armhf-3.15.4-20170202-1604.zip`
4. Download & install kali chroot via app

**Comment by AutinDroid** [comment]
@dverlik correct except you won't need to install chroot via nethunter app.

**Comment by dverlik** [comment]
@AutinDroid ...as I've already installed flashing the kali fs. Again, thank you a lot, this would be a real pain figuring out all by myself. Will try it and post results as people may be searching for a solution just like me.

**Comment by AutinDroid** [comment]
@dverlik actually, you install hammerheadmon-nougat not hammerhead-nougat

**Comment by dverlik** [comment]
@AutinDroid Got it.

**Comment by dverlik** [comment]
@AutinDroid Apparetly, after flashing `kernel-nethunter-hammerheadmon-nougat-3.15.4-20170202-1604.zip` as a secondary ROM it is impossible to boot into it, multirom reports an error and refuses to boot. For some reason the multirom_log.txt is unaccesible (every text manager just refuses to open it), so I can't provide more information until I am able to open it somehow.

Maybe it happens because the internal rom is a marshmallow with a flashed over it custom rootable kernel ?

**Comment by AutinDroid** [comment]
@dverlik If the kernel doesn't boot, I don't blame you. I earlier (about a week ago) created a kernel that was much better than the previous, using ElementalX. Unfortunately, after my pull request, the package has not been updated. The old kernel is really bad, some functions don't work, but the new one is much more stable. I can email you the updated package.

**Comment by dverlik** [comment]
@AutinDroid this would be even better. Also , it looks like my Nexus 5 does not fully support the nougat version of android, so I decided to fall back to the marshmallow version, as it has the `*mon` version too.

**Comment by AutinDroid** [comment]
@dverlik nexus 5 supports nougat, get a custom ROM like euclidean OS. The marshmallow Mon package is broken.

**Comment by dverlik** [comment]
@AutinDroid Now downloading euclideanOS with nougat kernel.

**Comment by akrolla** [comment]
Use the elementalx kernel u made @autindroid with the nethunerOS cm rom
@binkybear updated the gist for it like a week ago with an updated version.

**Comment by akrolla** [comment]
then you can have nougat

**Comment by akrolla** [comment]
https://gist.github.com/binkybear/8f0f4eee292d5263372f47c3c3edd08e

**Comment by akrolla** [comment]
dont forget to change the radio thats why its probably saying nexus 5 doesnt support it

**Comment by dverlik** [comment]
Okay, so the euclideanOS was stuck in bootloop.  I'll try the NethunterOS ROM.

**Comment by akrolla** [comment]
did it work?

**Comment by AutinDroid** [comment]
@dverlik the hammerheadmon package in the web is broken, I'll have to email you a fixed one.

**Comment by dverlik** [comment]
@akrolla Nope, it failed while installing.
```
Patching system image unconditionally...
E:unknown command [log]
E:unknown command [log]
E1001: failed to update system image.
E:unknown command [log]
Updater process ended with ERROR: 7
Failed to install ZIP!
```

**Comment by dverlik** [comment]
@AutinDroid It would be great. My mail is dverlik@2ch.hk

**Comment by akrolla** [comment]
And u followed all the instructions on the gist?

**Comment by dverlik** [comment]
@akrolla As the gist does not contain instructions for multirom install, I've created a new blank rom in multirom TWRP and tried to flash the `nethunteros-nethunter-hammerhead-cm-14.1-hammerhead.zip` to the blank rom.

**Comment by AutinDroid** [comment]
@dverlik  1.I have sent you an email with the kernel. 2. I understand that you probably value marshmallow for its availability in features Xposed and the fact you have personal data, but I think MultiRom is screwing things up. Probably best to back up your data and install a nougat ROM with the kernel and kalifs. I will try to update the kernel so you can use it as a main MultiRom ROM.

**Comment by dverlik** [comment]
@AutinDroid Ok, thanks. I'll try to install a fresh nougat ROM and post results.

**Comment by JulianVolodia** [comment]
How it's going?

**Comment by dverlik** [comment]
Last time I tried - it went into a bootloop. Going to  try to get it to work.

**Comment by mirzaatifbaig** [comment]
@AutinMitra Hi,I know this is straightforward and You don't know me But some of your Work intrigued me.I am talking about the hammerheadmon of Nethunter.You have submitted a zImage-dtb capable of Injection support inside chroot using chroot and A few patches.I think you know more than me that Quite a Few devices are made possible to bear monitor mode due to the Recent Nexmon Project.What I am trying to do is to enable monitor mode inside a Chrooted environment.The closest I have got to is to enable it on a note 3 inside a chroot using a patched kernel(Not so stable).But whenever I try to enable monitor mode using any script(Airodump-ng wifite anything inside the chroot)the Device freezes and I think the kernel goes into Panic.So,I have found that the hammerheadmon variant which comes with nethubterOS can do this without any kernel panic.Can you tell me the:

1-Toolchain you used

2-kernel source nougat

3-Patches

4-The problems you faced

5-The solution to them(If you want me to waste some time editing then no)

The goal I and a few colleagues of mine is to derive a method to get nethunterOS to every nexmon supported device.

I will wait for your Answers for 48 hours after which I will resend you a Message and maybe reach out from another platform

**Comment by AutinMitra** [comment]
@mirzaatifbaig
To be honest, I forgot about this project for a while, and don't remember much about how I did it.

1. I was using the androideabi toolchain straight from Google's git, I forgot which version
2. I believe I used ElementalX as a base for my kernel
3. I forgot exactly what the patches were (I don't even have the kernel source anymore), but if you some digging in binkybear's sources, you might find strike gold
4. I don't really recall many issues
5. Since I don't remember what I really did, I can't help with a solution
Sorry if I couldn't answer your question, hope this helps!
